### PR TITLE
feat: In-Game Pause

### PR DESCRIPTION
**Pull request by Piterga12**
The pause works well in every level (level 1 pause, level 2 pause, boss level pause).

(Whenever this is merged, the scenes will have to be updated (is as easy as adding the script PauseManager to the gameobject managers in the scene)

### CODE REVIEW

File: Source/DataModels/Timer/Timer.cpp [hunk 1]
```diff
@@ -17,12 +17,23 @@ float Timer::Read()
 		return static_cast<float>(startTick - stopTick);
 	}
 
-	return static_cast<float>(SDL_GetTicks() - startTick);
+	return static_cast<float>(SDL_GetTicks() - startTick + pauseTick);
+}
+
+void Timer::Pause()
+{
+	pauseTick = Read();
```
Comment (markmeliaa) [inline Source/DataModels/Timer/Timer.cpp:25]:
```suggestion
	pauseTick = static_cast<Uint32>(Read());
```

Comment (markmeliaa) [reply Source/DataModels/Timer/Timer.cpp:25]: Avoids raising a warning.

**Review by markmeliaa — COMMENTED** [review]
I don't have a controller on me right now so I can't try it, but code looks good enough!

File: Source/DataModels/Timer/Timer.cpp (hunk 1 above)
Comment (Piterga12) [reply Source/DataModels/Timer/Timer.cpp]: Ty, done!

File: Game/Scripts/FinalBossScript.cpp [hunk 2]
```diff
@@ -54,6 +54,10 @@ void FinalBossScript::Start()
 
 void FinalBossScript::Update(float deltaTime)
 {
+	if (isPaused)
+	{
+		return;
+	}
 	if (!target)
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/FinalBossScript.cpp:61]:
```suggestion
	if (!target || isPaused)
```
change it only if you find it easier to understand

File: Game/Scripts/PlayerManagerScript.cpp [hunk 3]
```diff
@@ -59,6 +62,11 @@ bool PlayerManagerScript::IsParalyzed() const
 	return movementManager->IsParalyzed();
 }
 
+bool PlayerManagerScript::GetIsPaused() const
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/PlayerManagerScript.cpp:65]:
```suggestion
bool PlayerManagerScript::IsPaused() const
```

File: Source/DataModels/Scene/Scene.cpp [hunk 4]
```diff
@@ -605,6 +605,24 @@ GameObject* Scene::SearchGameObjectByID(UID gameObjectID) const
 	return nullptr;
 }
 
+std::vector<GameObject*> Scene::SearchGameObjectByTag(const std::string& gameObjectTag) const
+{
+	std::vector<GameObject*> tagGameObjects;
+	for (int i = 0; GameObject* gameObject : sceneGameObjects)
```
Comment (MarcGonzalezMoratona) [inline Source/DataModels/Scene/Scene.cpp:611]:
```suggestion
	for (GameObject* gameObject : sceneGameObjects)
```
I think that you are not using this index

File: Source/DataModels/Scene/Scene.cpp [hunk 5]
```diff
@@ -605,6 +605,24 @@ GameObject* Scene::SearchGameObjectByID(UID gameObjectID) const
...
+	{
+		if (gameObject && gameObject->CompareTag(gameObjectTag))
+		{
+			tagGameObjects.push_back(gameObject);
+		}
+		i = i + 1;
```
Comment (MarcGonzalezMoratona) [inline Source/DataModels/Scene/Scene.cpp:617]:
```suggestion
```

File: Source/DataModels/Scene/Scene.cpp [hunk 6]
```diff
@@ -605,6 +605,24 @@ GameObject* Scene::SearchGameObjectByID(UID gameObjectID) const
...
+	}
+	if (tagGameObjects.size() == 0)
```
Comment (MarcGonzalezMoratona) [inline Source/DataModels/Scene/Scene.cpp:619]:
```suggestion
	if (tagGameObjects.empty())
```
https://cplusplus.com/reference/vector/vector/empty/

File: Source/DataModels/Scene/Scene.cpp (hunk 4 above)
Comment (Piterga12) [reply Source/DataModels/Scene/Scene.cpp:611]: True, thank you!

File: Game/Scripts/SwitchPlayerManagerScript.cpp [hunk 7]
```diff
@@ -78,10 +78,26 @@ void SwitchPlayerManagerScript::Start()
...
+	if (isPaused)
+	{
+		changePlayerTimer.Play();
+	}
+	
+	isPaused = false;
 	if (!isChangingPlayer)
 	{
 		if (input->GetKey(SDL_SCANCODE_C) != KeyState::IDLE && secondPlayer 
 			&& currentPlayer->GetComponent<PlayerManagerScript>()->IsGrounded()
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/SwitchPlayerManagerScript.cpp:99]: maybe we can store this in a variable
currentPlayer->GetComponent<PlayerManagerScript>()

File: Game/Scripts/UIGameManager.cpp [hunk 8]
```diff
@@ -92,13 +95,17 @@ void UIGameManager::MenuIsOpen()
 {
 	if (menuIsOpen)
 	{
+		manager->GetComponent<EnemiesManager>()->PauseEnemies(true);
 		mainMenuObject->Enable();
 		hudCanvasObject->Disable();
+		App->GetModule<ModulePlayer>()->GetPlayer()->GetComponent<PlayerManagerScript>()->FullPausePlayer(true);
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/UIGameManager.cpp:101]: at the end of the else-if block:
```suggestion
		App->GetModule<ModulePlayer>()->GetPlayer()->GetComponent<PlayerManagerScript>()->FullPausePlayer(menuIsOpen);
```

File: Game/Scripts/UIGameManager.cpp [hunk 9]
```diff
@@ -92,13 +95,17 @@ void UIGameManager::MenuIsOpen()
 {
 	if (menuIsOpen)
 	{
+		manager->GetComponent<EnemiesManager>()->PauseEnemies(true);
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/UIGameManager.cpp:98]: at the end of the if-else block:
```suggestion
		manager->GetComponent<EnemiesManager>()->PauseEnemies(menuIsOpen);
```

File: Source/DataModels/Scene/Scene.cpp (hunk 6 above)
Comment (Piterga12) [reply Source/DataModels/Scene/Scene.cpp:618]: True, that's better, thanks

File: Game/Scripts/EnemiesManager.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+class EnemiesManager : public Script
+{
+public:
+	EnemiesManager();
+	~EnemiesManager() override = default;
+
+	void PauseEnemies(bool paused);
+
+	const std::vector<GameObject*>& GetVectorEnemiesGameObjects() const;
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/EnemiesManager.h:19]:
```suggestion
	const std::vector<GameObject*>& GetEnemiesGameObjects() const;
```

File: Game/Scripts/EnemiesManager.h [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+	void SetVectorEnemiesGameObjects(const std::vector<GameObject*>& enemiesGameObjects);
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/EnemiesManager.h:20]:
```suggestion
	void SetEnemiesGameObjects(const std::vector<GameObject*>& enemiesGameObjects);
```

File: Game/Scripts/EnemiesManager.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+{
+	REGISTER_FIELD(enemyTag, std::string);
+	REGISTER_FIELD(bulletTag, std::string);
+}
+
+void EnemiesManager::Start()
+{
+	scene = App->GetModule<ModuleScene>()->GetLoadedScene();
+	enemiesGameObjects = scene->SearchGameObjectByTag(enemyTag);
+
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/EnemiesManager.cpp:28]:
```suggestion
```

File: Game/Scripts/EnemiesManager.cpp [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	
+	for (int i = 0; i < alluraBulletGameObjects.size(); i++)
+	{
+
+		alluraBulletGameObjects[i]->GetComponent<LightAttackBullet>()->SetPauseBullet(paused);
+		
+	}
+}
+
+const std::vector<GameObject*>& EnemiesManager::GetVectorEnemiesGameObjects() const
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/EnemiesManager.cpp:64]: this method returns the address to the first element of the vector, is this intended?
pay attention to the & after the vector returned

File: Game/Scripts/EnemiesManager.h (hunk 10 above)
Comment (Piterga12) [reply Game/Scripts/EnemiesManager.h:19]: Actually this function wasn't used in the end, so can be removed, so i will, thanks

**Comment by Piterga12** [comment]
Now the pause is applied to the boss and some other things that were missed, please check that the pause works well in every level (level 1 pause, level 2 pause, boss level pause).

(Whenever this is merged, the scenes will have to be updated (is as easy as adding the script PauseManager to the gameobject managers in the scene)

**Comment by diegusmejia117** [comment]
Great work, but just 2 things: 

Level 2:
-when pausing while hacking, allura becomes deparalyzed but can still hack.
- when pausing while a venomite is the elevator in the shield fase of the miniboss the venomite becomes stuck flying in the position it paused in.

image

**Comment by markmeliaa** [comment]
> Great work, but just 2 things:
> 
> Level 2: -when pausing while hacking, allura becomes deparalyzed but can still hack.
> 
> * when pausing while a venomite is the elevator in the shield fase of the miniboss the venomite becomes stuck flying in the position it paused in.
> 
> image

Maybe @DADSILVER may have an idea why that's happening. Seems like something is not set back correctly after the game is resumed from being paused.

File: Game/Scripts/PlayerMoveScript.cpp [hunk 14]
```diff
@@ -79,6 +79,8 @@ void PlayerMoveScript::PreUpdate(float deltaTime)
 		{
 			componentAnimation->SetParameter("IsRunning", false);
 			componentAnimation->SetParameter("IsDashing", false);
+			owner->GetComponent<ComponentRigidBody>()->GetRigidBody()->setLinearVelocity(btVector3(0.f,
+				owner->GetComponent<ComponentRigidBody>()->GetRigidBody()->getLinearVelocity().getY(), 0.f));
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/PlayerMoveScript.cpp:83]: you could store the linearVelocity in a variable if you want to make this line shorter, it's up to you!

File: Game/Scripts/EnemyClass.cpp [hunk 15]
```diff
@@ -71,6 +71,10 @@ EnemyTypes EnemyClass::GetEnemyType() const
 	return enemyType;
 }
 
+void EnemyClass::PauseEnemy(bool isPaused)
+{
+	this->isPaused = isPaused;
+}
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/EnemyClass.cpp:77]:
```suggestion
}

```

File: Game/Scripts/BossChargeRockScript.h [hunk 16]
```diff
@@ -42,4 +44,8 @@ class BossChargeRockScript : public Script
 	// Modifiable values
 	float fallingRockDamage;
 	float despawnMaxTimer;
+
+	bool isPaused;
+	bool isUnpaused;
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/BossChargeRockScript.h:49]: why do we need isPaused and isUnpaused? Can't we just make !isPaused when needed?

File: Game/Scripts/EnemyVenomiteScript.cpp [hunk 17]
```diff
@@ -86,6 +94,7 @@ void EnemyVenomiteScript::Update(float deltaTime)
 		}
 	}
 
+
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/EnemyVenomiteScript.cpp:97]:
```suggestion
```

File: Game/Scripts/SwitchPlayerManagerScript.cpp [hunk 18]
```diff
@@ -83,11 +83,27 @@ void SwitchPlayerManagerScript::Start()
...
+		if (!isPaused)
+		{
+			changePlayerTimer.Pause();
+		}
+		isPaused = true;
+		return;
+	}
+	if (isPaused)
+	{
+		changePlayerTimer.Play();
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/SwitchPlayerManagerScript.cpp:97]: Maybe I don't understand it well but I see that here the timer plays when the game is paused?

File: Game/Scripts/FinalBossScript.cpp [hunk 19]
```diff
@@ -48,11 +48,39 @@ void FinalBossScript::Start()
...
+
+		bossPhase = FinalBossPhases::NEUTRAL;
+		bossState = FinalBossStates::IDLE;
+
+		rigidBody->GetRigidBody()->setLinearVelocity(btVector3(0, 0, 0));
+		rigidBody->GetRigidBody()->setGravity(btVector3(0, 0, 0));
+		paused = true;
+		return;
+	}
+	if (paused)
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/FinalBossScript.cpp:74]: why do you need a paused boolean if there's a isPaused one?

File: Game/Scripts/PauseManager.cpp [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+	PauseEnemies(paused);
+	PauseBullets(paused);
+	PauseRocks(paused);
+	App->GetModule<ModulePlayer>()->GetPlayer()->GetComponent<PlayerManagerScript>()->FullPausePlayer(paused);
+}
+
+void PauseManager::PauseEnemies(bool paused)
+{
+	for (int i = 0; i < enemiesGameObjects.size(); i++)
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/PauseManager.cpp:52]:
```suggestion
	for (int i = 0; i < enemiesGameObjects.size(); ++i)
```

File: Game/Scripts/PauseManager.cpp [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+		else
+		{
+			enemiesGameObjects[i]->GetComponent<ComponentAnimation>()->Enable();
+		}
+	}
+}
+
+void PauseManager::PauseBullets(bool paused)
+{
+	for (int i = 0; i < bulletGameObjects.size(); i++)
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/PauseManager.cpp:68]:
```suggestion
	for (int i = 0; i < bulletGameObjects.size(); ++i)
```

File: Game/Scripts/PauseManager.cpp [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+		{
+			bulletGameObjects[i]->GetComponent<RangedFastAttackBullet>()->SetPauseBullet(paused);
+		}
+		else if (bulletGameObjects[i]->IsEnabled())
+		{
+			bulletGameObjects[i]->GetComponent<BossMissilesMissileScript>()->SetIsPaused(paused);
+		}
+	}
+	
+	for (int i = 0; i < alluraBulletGameObjects.size(); i++)
```
Comment (MarcGonzalezMoratona) [inline Game/Scripts/PauseManager.cpp:80]:
```suggestion
	for (int i = 0; i < alluraBulletGameObjects.size(); ++i)
```

File: Game/Scripts/SwitchPlayerManagerScript.cpp (hunk 18 above)
Comment (Piterga12) [reply Game/Scripts/SwitchPlayerManagerScript.cpp:97]: I will rename the bool to isUnpaused, as we need to unpause what its inside as if i don't do it just once, we get some issues

File: Game/Scripts/FinalBossScript.cpp (hunk 19 above)
Comment (Piterga12) [reply Game/Scripts/FinalBossScript.cpp:74]: I will rename the bool to isUnpaused, as we need to unpause what its inside as if i don't do it just once, we get some issues

**Comment by Piterga12** [comment]
> Great work, but just 2 things:
> 
> Level 2: -when pausing while hacking, allura becomes deparalyzed but can still hack.
> 
> * when pausing while a venomite is the elevator in the shield fase of the miniboss the venomite becomes stuck flying in the position it paused in.
> 
> image

Both issues are fixed now, please test if you find anything else, thanks!

**Review by AuxPort66 — APPROVED** [review]

**Review by AuxPort66 — APPROVED** [review]

**Review by Vlukkee — APPROVED** [review]

**Review by AuxPort66 — APPROVED** [review]

**Pull request closed without merging**